metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/MetricQueryBuilder.java

```diff
@@ -0,0 +1,278 @@
+import org.hypertrace.core.attribute.service.client.AttributeServiceClient;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadata;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadataFilter;
+import org.hypertrace.core.query.service.api.ColumnIdentifier;
+import org.hypertrace.core.query.service.api.Expression;
+import org.hypertrace.core.query.service.api.Function;
+import org.hypertrace.core.query.service.api.LiteralConstant;
+import org.hypertrace.core.query.service.api.Operator;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.hypertrace.core.query.service.api.Value;
+import org.hypertrace.core.query.service.api.ValueType;
+import org.hypertrace.core.query.service.client.QueryServiceClient;
+import org.hypertrace.gateway.service.v1.common.FunctionType;
+
+class MetricQueryBuilder {
+
+  private static final String START_TIME_ATTRIBUTE_KEY = "startTime";
+  private static final String DATE_TIME_CONVERTER = "dateTimeConvert";
+  private static final StringJoiner dotJoiner = new StringJoiner(".");
+
+  private final AttributeServiceClient attributesServiceClient;
+
+  MetricQueryBuilder(AttributeServiceClient attributesServiceClient) {
+    this.attributesServiceClient = attributesServiceClient;
+  }
+
+  QueryRequest buildMetricQueryRequest(
+      MetricSelection metricSelection, long startTime, long endTime, String tenantId) {
+    QueryRequest.Builder builder = QueryRequest.newBuilder();
+    String timeColumn =
+        getTimestampAttributeId(tenantId, metricSelection.getMetricAttribute().getScope());
+    if (null == timeColumn) {
+      throw new IllegalArgumentException("Error time column is null");
+    }
+    org.hypertrace.core.query.service.api.Filter filter =
+        convertFilter(metricSelection.getFilter());
+    builder.setFilter(addTimeFilter(filter, startTime, endTime, timeColumn));
+
+    setSelection(
+        builder,
+        metricSelection.getMetricAttribute(),
+        metricSelection.getMetricAggregationFunction());
+
+    builder.addGroupBy(
+        createTimeColumnGroupByExpression(
+            timeColumn, isoDurationToSeconds(metricSelection.getMetricAggregationInterval())));
+    builder.setLimit(QueryServiceClient.DEFAULT_QUERY_SERVICE_GROUP_BY_LIMIT);
+
+    return builder.build();
+  }
+
+  static long isoDurationToSeconds(String duration) {
+    Duration d = java.time.Duration.parse(duration);
+    return d.get(ChronoUnit.SECONDS);
+  }
+
+  static org.hypertrace.core.query.service.api.Filter convertFilter(Filter filter) {
+    if (filter.equals(Filter.getDefaultInstance())) {
+      return org.hypertrace.core.query.service.api.Filter.getDefaultInstance();
+    }
+    switch (filter.getFilterCase()) {
+      case LEAF_FILTER:
+        return convertLeafFilter(filter.getLeafFilter());
+      case COMPOSITE_FILTER:
+        return convertCompositeFilter(filter.getCompositeFilter());
+      default:
+        return null;
```

Comment:
will there be an exception if it falls in this case?

Reply:
yes when the filter case is `unspecified`